### PR TITLE
[NU-1223] Tapir nodes endpoint

### PR DESCRIPTION
**Pull request by ForrestFairy**
## Describe your changes
Replacing NodesResources with NodesApiEndpoints and NodesApiHttpService which are using Tapir.

The most challenging part is a way to make `TypingResult` static for Tapir encoders and decoders. We get to know what class of variable is given once the request is made which makes it hard to parse it correctly.

Places where `TypingResult` is sent either way we will pass a Json. On the server side it will be decoded and used as a normal `TypingResult` and before sending response it will be encoded back to Json.

## Checklist before merge
- [x] Related issue ID is placed at the beginning of PR title in \[brackets\] (can be GH issue or Nu Jira issue)
- [ ] Code is cleaned from temporary changes and commented out lines
- [ ] Parts of the code that are not easy to understand are documented in the code
- [ ] Changes are covered by automated tests
- [ ] Showcase in dev-application.conf added to demonstrate the feature
- [ ] Documentation added or updated
- [ ] Added entry in _Changelog.md_ describing the change from the perspective of a public distribution user
- [ ] Added _MigrationGuide.md_ entry in the appropriate subcategory if introducing a breaking change
- [x] Verify that PR will be squashed during merge

### CODE REVIEW

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/api/TypeDtoEncoders.scala [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+      case typingDto.TypedNullDto        => encodeNull
+      case TypedUnionDto(classes) =>
+        JsonObject("union" -> fromValues(classes.map(typ => fromJsonObject(encodeTypingResultDto(typ))).toList))
+    })
+      .+:(typeField -> fromString(TypingTypeDto.forType(dto).toString))
+      .+:("display" -> fromString(dto.display))
+
+  private def encodeSingleTypingResult(dto: SingleTypingResultDto): JsonObject =
+    dto match {
+      case TypedObjectTypingResultDto(fields, objType, additionalInfo) =>
```
Comment (lukasz-bigorajski) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/api/TypeDtoEncoders.scala:40]: Cannot we have separate encoders in such way:
```scala
case TypedObjectTypingResultDto(fields, objType, additionalInfo) =>
  encodeTypedObjectTypingResultDto...
case TypedDictDto(dictId, valueType) =>
  encodeTypedDictDto
```
?

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/api/NodesApiEndpoints.scala [hunk 2]
```diff
@@ -0,0 +1,530 @@
...
+                toTypingResult(typ)
+              )
+            )
+          case typedObjectWithValue: TypedObjectWithValueDto =>
+            val underlying = Typed.genericTypeClass(
+              modelData.modelClassLoader.classLoader.loadClass(typedObjectWithValue.waitRefClazzName),
+              List.empty
+            )
+            val trueValue = SimpleObjectEncoder.decode(underlying, typedObjectWithValue.waitValue.hcursor) match {
+              case Left(value)  => ???
```
Comment (lukasz-bigorajski) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/api/NodesApiEndpoints.scala:221]: What with this?

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/api/TypeDtoEncoders.scala [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+
+  private def encodeTypingResultDto(dto: TypingResultDto): JsonObject =
+    (dto match {
+      case single: SingleTypingResultDto => encodeSingleTypingResult(single)
+      case typingDto.UnknownDto          => encodeUnknown
+      case typingDto.TypedNullDto        => encodeNull
+      case TypedUnionDto(classes) =>
+        JsonObject("union" -> fromValues(classes.map(typ => fromJsonObject(encodeTypingResultDto(typ))).toList))
+    })
+      .+:(typeField -> fromString(TypingTypeDto.forType(dto).toString))
```
[the remark this answers is not in the document]
Comment (ForrestFairy) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/api/TypeDtoEncoders.scala]: Most of the code for this file was copied from `TypeEncoders.scala`. And as only parts were needed to be changed for static encoders and decoders to work, I have left the rest unchanged. For example this part

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/api/typingDto.scala [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+  }
+
+  case class TypedObjectWithValueDto private[typingDto] (
+      underlying: TypedClassDto,
+      value: Any,
+      waitRefClazzName: String = "",
+      waitValue: Json = Json.fromString("")
+  ) extends TypedObjectWithDataDto {
+    val maxDataDisplaySize: Int         = 15
+    val maxDataDisplaySizeWithDots: Int = maxDataDisplaySize - "...".length
```
[the remark this answers is not in the document]
Comment (ForrestFairy) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/api/typingDto.scala]: I understand it as "max amount of data displayed when dots are needed" (as it's longer then 15) but this part was also copied from `typing.scala` so that's only my thoughts of what it means

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/api/typingDto.scala [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+      TypedTaggedValueDto(SingleTypingResultDto.apply(typedTaggedValue.underlying), typedTaggedValue.tag)
+    }
+
+  }
+
+  case class TypedObjectWithValueDto private[typingDto] (
+      underlying: TypedClassDto,
+      value: Any,
+      waitRefClazzName: String = "",
+      waitValue: Json = Json.fromString("")
```
[the remark this answers is not in the document]
Comment (ForrestFairy) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/api/typingDto.scala]: That's the main problem in this task. I have not found any other way to be able to parse `Any` value and to be able to have static encoder, decoder and schema for it which are required by Tapir. So my current solution is to parse Json, on the server side parse it to `TypedClass` which has only `value: Any` and no Json, process the request and still on server side parse this `value: Any` to Json and send back result with Json

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/services/NodesApiHttpService.scala [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+      }
+  }
+
+  expose {
+    nodesApiEndpoints.parametersValidationEndpoint
+      .serverSecurityLogic(authorizeKnownUser[String])
+      .serverLogic { implicit loggedUser => pair =>
+        val (processingType, request) = pair
+
+        try {
```
[the remark this answers is not in the document]
Comment (ForrestFairy) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/services/NodesApiHttpService.scala]: I have tried to extract it, but I'm not sure if I have done it correctly. Also only the first line is the same in this try/catch so I didn't know if it's possible here to extract more

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/api/NodesApiEndpoints.scala [hunk 7]
```diff
@@ -0,0 +1,530 @@
...
+object NodesApiEndpoints {
+
+  object Dtos {
+
+    implicit val typeDtoEncoderAsObject: Encoder.AsObject[TypingResultDto] = TypeDtoEncoders.typingResultEncoder
+    implicit val typeDtoEncoders: Encoder[TypingResultDto] = TypeDtoEncoders.typingResultEncoder.mapJson(json => json)
+    private val typingDtoDecoder                           = new TypingResultDtoDecoder()
+    implicit val typingResultDtoDecoder: Decoder[TypingResultDto] = typingDtoDecoder.decodeTypingResultDto
+
+    object ProcessNameCodec {
```
[the remark this answers is not in the document]
Comment (ForrestFairy) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/api/NodesApiEndpoints.scala]: I see you have created `TapirCodecs` in your pull request, should I then leave this until you merge your changes or should I do the same thing and in the future just resolve conflict between two versions?

[the remark this answers is not in the document]
Comment (lukasz-bigorajski) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/api/NodesApiEndpoints.scala]: I prefer to do the same thing, and then the person who merges later need to resolve conflicts.

File: components-api/src/main/scala/pl/touk/nussknacker/engine/api/typed/typing.scala [hunk 8]
```diff
@@ -151,7 +151,7 @@ object typing {
 
   // It is not a case class because we want to ignore the order of elements but still ensure that it has >= 2 elements
   // Because of that, we have our own equals and hashCode
-  class TypedUnion private[typing] (
```
Comment (lukasz-bigorajski) [inline components-api/src/main/scala/pl/touk/nussknacker/engine/api/typed/typing.scala:154]: Shouldn't it stay private?

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/api/NodesApiEndpoints.scala [hunk 9]
```diff
@@ -0,0 +1,497 @@
...
+        nodeData: NodeData,
+        processProperties: ProcessProperties,
+        variableTypes: Map[String, TypingResultInJson],
+        branchVariableTypes: Option[Map[String, Map[String, TypingResultInJson]]],
+        outgoingEdges: Option[List[Edge]]
+    )
+
+    implicit val nodeValidationRequestDtoEmptyEncoder: Encoder[NodeValidationRequestDto] =
+      Encoder.encodeJson.contramap[NodeValidationRequestDto](_ => Json.Null)
+//    implicit val nodeValidationRequestDtoEmptyEncoder: Encoder[NodeValidationRequestDto] = ???
```
Comment (lukasz-bigorajski) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/api/NodesApiEndpoints.scala:210]: I would remove this commented lines

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/api/NodesApiEndpoints.scala [hunk 10]
```diff
@@ -0,0 +1,497 @@
...
+    @derive(decoder, schema)
+    final case class NodeValidationRequestDto(
+        nodeData: NodeData,
+        processProperties: ProcessProperties,
+        variableTypes: Map[String, TypingResultInJson],
+        branchVariableTypes: Option[Map[String, Map[String, TypingResultInJson]]],
+        outgoingEdges: Option[List[Edge]]
+    )
+
+    implicit val nodeValidationRequestDtoEmptyEncoder: Encoder[NodeValidationRequestDto] =
```
Comment (lukasz-bigorajski) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/api/NodesApiEndpoints.scala:208]: Can we move the encoder to the companion object?

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/api/NodesApiEndpoints.scala [hunk 11]
```diff
@@ -0,0 +1,497 @@
...
+    final case class NodeValidationResultDto(
+        parameters: Option[List[UIParameterDto]],
+        expressionType: Option[TypingResult],
+        validationErrors: List[NodeValidationError],
+        validationPerformed: Boolean
+    )
+
+    implicit val nodeValidationRequestDtoDecoder: Decoder[NodeValidationResultDto] =
+      Decoder.instance[NodeValidationResultDto](_ => ???)
+//    implicit val nodeValidationRequestDtoDecoder: Decoder[NodeValidationResultDto] = ???
```
Comment (lukasz-bigorajski) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/api/NodesApiEndpoints.scala:228]: Can we remove the commented line?

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/api/NodesApiEndpoints.scala [hunk 12]
```diff
@@ -0,0 +1,497 @@
...
+    final case class ExpressionSuggestionRequestDto(
+        expression: Expression,
+        caretPosition2d: CaretPosition2d,
+        variableTypes: Map[String, TypingResultInJson]
+    ) {
+
+      def decodedVariableTypes(decoder: Decoder[TypingResult]): Map[String, TypingResult] =
+        variableTypes.map { case (key, result) =>
+          (key, decoder.decodeJson(result).getOrElse(Unknown))
+        } // todo unknown for now
```
Comment (lukasz-bigorajski) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/api/NodesApiEndpoints.scala:323]: Is it ok to Uknown TypingResult to stay? Or should we throw some exception instead?

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/api/NodesApiEndpoints.scala [hunk 13]
```diff
@@ -0,0 +1,497 @@
...
+    }
+
+    def prepareTestFromParametersDecoder(modelData: ModelData): Decoder[TestFromParametersRequest] = {
+      implicit val typeDecoder: Decoder[TypingResult] = prepareTypingResultDecoder(modelData)
+      implicit val testSourceParametersDecoder: Decoder[TestSourceParameters] =
+        deriveConfiguredDecoder[TestSourceParameters]
+      deriveConfiguredDecoder[TestFromParametersRequest]
+    }
+
+    //    These 3 doesn't look as used anywhere
```
Comment (lukasz-bigorajski) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/api/NodesApiEndpoints.scala:366]: If it is not being used anywhere I would remove it

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/api/NodesApiEndpoints.scala [hunk 14]
```diff
@@ -0,0 +1,497 @@
...
+      sourceId: String,
+      parameterExpressions: Map[String, Expression]
+  )
+
+  @JsonCodec(encodeOnly = true) final case class TestFromParametersRequest(
+      sourceParameters: TestSourceParameters,
+      scenarioGraph: ScenarioGraph
+  )
+
+  @JsonCodec(encodeOnly = true) final case class ParametersValidationResult(
```
Comment (lukasz-bigorajski) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/api/NodesApiEndpoints.scala:396]: Is it used anywhere?

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/api/NodesApiEndpoints.scala [hunk 15]
```diff
@@ -0,0 +1,497 @@
...
+    def apply(
+        request: ParametersValidationRequestDto
+    )(typingResultDecoder: Decoder[TypingResult]): ParametersValidationRequest = {
+      new ParametersValidationRequest(
+        request.parameters.map { parameter =>
+          UIValueParameter(
+            name = parameter.name,
+            typ = typingResultDecoder
+              .decodeJson(parameter.typ)
+              .getOrElse(Unknown), // todo for now is Unknown, maybe ex should be rethrown
```
Comment (lukasz-bigorajski) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/api/NodesApiEndpoints.scala:418]: What with this todo?

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/api/NodesApiEndpoints.scala [hunk 16]
```diff
@@ -0,0 +1,497 @@
...
+            expression = parameter.expression
+          )
+        },
+        request.variableTypes.map { case (key, typDto) =>
+          (key, typingResultDecoder.decodeJson(typDto).getOrElse(Unknown)) // todo the same as higher todo
```
Comment (lukasz-bigorajski) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/api/NodesApiEndpoints.scala:423]: The same here

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/api/NodesApiEndpoints.scala [hunk 17]
```diff
@@ -0,0 +1,497 @@
...
+  )
+
+  object NodeValidationRequest {
+
+    def apply(node: NodeValidationRequestDto)(typingResultDecoder: Decoder[TypingResult]): NodeValidationRequest = {
+      new NodeValidationRequest(
+        nodeData = node.nodeData,
+        processProperties = node.processProperties,
+        variableTypes = node.variableTypes.map { case (key, typingResult) =>
+          (key, typingResultDecoder.decodeJson(typingResult).getOrElse(Unknown)) // todo else -> unknown for now
```
Comment (lukasz-bigorajski) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/api/NodesApiEndpoints.scala:461]: The same here

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/api/NodesApiEndpoints.scala [hunk 18]
```diff
@@ -0,0 +1,497 @@
...
+      name: ProcessName
+  )
+
+  final case class ExpressionSuggestionRequest(
+      expression: Expression,
+      caretPosition2d: CaretPosition2d,
+      variableTypes: Map[String, TypingResult]
+  )
+
+  object ExpressionSuggestionRequest {
```
Comment (lukasz-bigorajski) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/api/NodesApiEndpoints.scala:489]: Is it used anywhere?

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/api/TypeDtoEncoders.scala [hunk 19]
```diff
@@ -0,0 +1,195 @@
...
+//import io.circe._
+//import io.circe.Json._
+//import pl.touk.nussknacker.engine.api.typed.typing
+//import pl.touk.nussknacker.engine.api.typed.typing.AdditionalDataValue
+//import pl.touk.nussknacker.engine.util.Implicits.RichScalaMap
+//import pl.touk.nussknacker.ui.api.TypeDtoEncoders.typeField
+//import pl.touk.nussknacker.ui.api.TypingTypeDto.TypingType
+//import pl.touk.nussknacker.ui.api.typingDtos._
+//
+//object TypeDtoEncoders {
```
Comment (lukasz-bigorajski) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/api/TypeDtoEncoders.scala:12]: We should remove this file :)

**Review by lukasz-bigorajski — CHANGES_REQUESTED** [review]

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/api/NodesApiEndpoints.scala [hunk 20]
```diff
@@ -0,0 +1,534 @@
...
+        sourceParameters: TestSourceParameters,
+        scenarioGraph: ScenarioGraph
+    )
+
+    @JsonCodec(encodeOnly = true) final case class ParametersValidationRequest(
+        parameters: List[UIValueParameter],
+        variableTypes: Map[String, TypingResult]
+    )
+
+    object ParametersValidationRequest {
```
Comment (lukasz-bigorajski) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/api/NodesApiEndpoints.scala:418]: I would move this logic to the HttpService because it has more functions than a simple dto.

**Review by lukasz-bigorajski — APPROVED** [review]

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/api/typingDtoSchemas.scala [hunk 21]
```diff
@@ -0,0 +1,147 @@
...
+object typingDtoSchemas {
+
+  implicit lazy val typingResult: Schema[TypingResult] = Schema.derived
+
+  implicit lazy val singleTypingResultSchema: Schema[SingleTypingResult] =
+    Schema.derived.hidden(true)
+
+  implicit lazy val additionalDataValueSchema: Schema[AdditionalDataValue] = Schema.derived
+
+  implicit lazy val typedObjectTypingResultSchema: Schema[TypedObjectTypingResult] = {
```
[the remark this answers is not in the document]
Comment (ForrestFairy) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/api/typingDtoSchemas.scala]: I don't see a way to attach this schemas to codecs to be automatically informed about a change. They couldn't be derived due to the dynamic nature of it. Maybe some tests will fail if the codecs are changed.
But I would say there is currently a probability of codecs and schemas desynchronizing without our knowledge. I can add some comment in `typing.scala` to decrease this probability, but I still wouldn't be 100% sure

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/services/NodesApiHttpService.scala [hunk 22]
```diff
@@ -0,0 +1,327 @@
...
+  private def toExpressionSuggestionResponse(suggestions: Future[List[ExpressionSuggestion]]) = {
+    Future[Either[NodesError, List[ExpressionSuggestionDto]]](
+      suggestions.value match {
+        case Some(value) =>
+          value match {
+            case Failure(exception) => Left(MalformedTypingResult(exception.getMessage))
+            case Success(list)      => Right(list.map { expression => ExpressionSuggestionDto(expression) })
+          }
+        case None => // Request not completed yet
+          throw new IllegalStateException()
```
Comment (mk-software-pl) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/services/NodesApiHttpService.scala:250]: message needed (explaining what is the illegal state here)

**Review by mk-software-pl — APPROVED** [review]
LGTM
